### PR TITLE
Submit gather to deadline

### PR DESCRIPTION
**Pull request by jm22dogs**
## Brief description
Implementing publishing in farm for the gather process

## Description
This is a W.I.P. pull request, as it is not fully tested and it is not fully working.
This PR is created so that Max can look at the code.

The idea is to set up a standard for plugins that can allow publishing directly to the farm. The bulk of the code resides in the module `openpype/plugins/publish/submit_gather_deadline.py`.

### Minor changes done in existing modules:

1. `openpype/modules/deadline/plugins/publish/submit_publish_job.py`: extending families and hosts in which this plugin is activated (tray and gather.farm) **and** adding an exception at the construction of the publish deadline job for when there is not a dependent job previous to the publish job one and an `_id` cannot be found.
2. `openpype/plugins/publish/collect_gather_data.py`: deactivating the context key "farm" so that the `/openpype/plugins/publish/integrate.py` module extends the instance environment with the gather files, otherwise this plugin won't have extended it and data would be missing. The `farm` key is set to `True` in the `openpype/plugins/publish/submit_gather_deadline.py` module.
3. `openpype/modules/deadline/plugins/publish/validate_expected_and_rendered_files.py`:  this plugin checks i/o of rendered files. It generaly gets the framerange from the previous job (the job that does the render), in the case of gather, there may not be previous job. If this is the case, this plugin fails as there is no preivous job data, so now it has a fix for whenever there is no dependent job; for now and for debuging purposes this fix only acts if gather is in the name of the instance.data otherwise it will continue the usual logic. Adding also some log.debugs.

## Testing notes
1. The `Gather instances on farm` toggle must be activated in the traypublisher interface.
Expected behaviour: the local publish won't happen and a new job should appear in Deadline.
2. At the moment it does dual publishing, as the local publish doesn't stop even if the deadline publish is on.

## TODO
1. `validate_expected_and_rendered_files.py`: gather frame range should be calculated from the repres files, and fallback to frameStartHandle - frameEndHandle for when the gather is on videos
2. `submit_gather_deadline.py`: look at why we would use `write` instead of `gather.write` or whatever in the method `redefine_families`
3. Thumbnail is not being created with the deadline publish
4. The local publish is not being interrupted after deadline publish
5. The version is not in ftrack's folder `root/dataio`

### CODE REVIEW

File: openpype/modules/deadline/plugins/publish/validate_expected_and_rendered_files.py [hunk 1]
```diff
@@ -21,7 +21,12 @@ class ValidateExpectedFiles(pyblish.api.InstancePlugin):
 
     def process(self, instance):
         self.instance = instance
-        frame_list = self._get_frame_list(instance.data["render_job_id"])
+
+        if "gather" in instance.data["name"]:
+            frame_range = range(1001, int(instance.data["frameEndHandle"]))
```
Comment (maxpareschi) [inline openpype/modules/deadline/plugins/publish/validate_expected_and_rendered_files.py:26]: this is hardwired at 1001, but we need to actually get the frame number from the first and last frame of the repre. even frameEndHandle is not always present. it's probably best to take the first and last items from repre["files"] if it's a list and regex to find a number sequence from the end of the file string. If instead this is a mov (so repre["files"] is just a string we can fallback to frameStartHandle and frameEndHandle, and if they are not in the instance we need to do (frameEnd + handleEnd) - (frameStart - handleStart)

File: openpype/plugins/publish/collect_gather_data.py [hunk 2]
```diff
@@ -74,7 +74,14 @@ def process(self, instance):
                 "gather.farm",
                 "publish_on_farm"
             ])
-            instance.data["farm"] = True
+
+
+            instance.data["farm"] = False # Juan: if we set this to true, the integrate.py
```
Comment (maxpareschi) [inline openpype/plugins/publish/collect_gather_data.py:79]: ok!

File: openpype/plugins/publish/submit_gather_deadline.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        it is not recogniced by following plugins and thus they are turned off."""
+
+        if "render.farm" in families:
+            instance.data['family'] = 'write'
+            families.insert(0, "render2d")
+        elif "prerender.farm" in families:
+            instance.data['family'] = 'write'
+            families.insert(0, "prerender")
+        elif "gather.farm" in families:
+            instance.data['family'] = 'write'
```
Comment (maxpareschi) [inline openpype/plugins/publish/submit_gather_deadline.py:123]: mmh but if this is going to be renamed as write how can we intercept the gather.farm and turn it into a normal gather in the family afterwards?

**Review by maxpareschi — CHANGES_REQUESTED** [review]

File: openpype/plugins/publish/submit_gather_deadline.py (hunk 3 above)
Comment (jm22dogs) [reply openpype/plugins/publish/submit_gather_deadline.py:123]: For now I just left it like that to replicate what the OP team did with Nuke; I don't know whether we actually need to turn this back into gather, is there a reason why? the publish seems OK if I am not mistaken.
As there are many moving pieces here, after I am done with the parts I am currently testing I will revisit this then.

File: openpype/modules/deadline/plugins/publish/validate_expected_and_rendered_files.py (hunk 1 above)
Comment (jm22dogs) [reply openpype/modules/deadline/plugins/publish/validate_expected_and_rendered_files.py:26]: perfect, I hardcoded 1001 for debugging purposes, I will add what you suggested here after I'm done with the parts I am looking at now.

File: openpype/plugins/publish/submit_gather_deadline.py (hunk 3 above)
Comment (maxpareschi) [reply openpype/plugins/publish/submit_gather_deadline.py:123]: well, the actual part of this is that i fear that the templated transcodes will not get picked up with the correct profiles. There is a nuke profile for render, but we should not operate under the assumption that it's the same as the gather.
Therefore, what you did in getting the family as write here works, but then it gets switched as "render", not gather, which can have a different profile.

Comment (maxpareschi) [reply openpype/plugins/publish/submit_gather_deadline.py:123]: tested on your logs and it does actually get converted to "review" family. We need to convert to gather to make this work consistently

**Comment by maxpareschi** [comment]
I'm going to merge this tentatively and revert if i can't find a way to make it work

**Pull request merged**